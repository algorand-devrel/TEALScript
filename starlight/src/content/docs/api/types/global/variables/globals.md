---
editUrl: false
next: false
prev: false
title: "globals"
---

> **`const`** **globals**: `Object`

## Type declaration

### assetCreateMinBalance

> **assetCreateMinBalance**: [`uint64`](../type-aliases/uint64.md)

### assetOptInMinBalance

> **assetOptInMinBalance**: [`uint64`](../type-aliases/uint64.md)

### callerApplicationAddress

> **callerApplicationAddress**: [`Address`](../classes/Address.md)

### callerApplicationID

> **callerApplicationID**: [`Application`](../classes/Application.md)

### creatorAddress

> **creatorAddress**: [`Address`](../classes/Address.md)

### currentApplicationAddress

> **currentApplicationAddress**: [`Address`](../classes/Address.md)

### currentApplicationID

> **currentApplicationID**: [`Application`](../classes/Application.md)

### genesisHash

> **genesisHash**: [`bytes32`](../type-aliases/bytes32.md)

### groupID

> **groupID**: [`bytes32`](../type-aliases/bytes32.md)

### groupSize

> **groupSize**: [`uint64`](../type-aliases/uint64.md)

### latestTimestamp

> **latestTimestamp**: [`uint64`](../type-aliases/uint64.md)

### logicSigVersion

> **logicSigVersion**: [`uint64`](../type-aliases/uint64.md)

### maxTxnLife

> **maxTxnLife**: [`uint64`](../type-aliases/uint64.md)

### minBalance

> **minBalance**: [`uint64`](../type-aliases/uint64.md)

### minTxnFee

> **minTxnFee**: [`uint64`](../type-aliases/uint64.md)

### opcodeBudget

> **opcodeBudget**: [`uint64`](../type-aliases/uint64.md)

### round

> **round**: [`uint64`](../type-aliases/uint64.md)

### zeroAddress

> **zeroAddress**: [`Address`](../classes/Address.md)

## Source

[types/global.d.ts:800](https://github.com/algorandfoundation/tealscript/blob/18ba30a9/types/global.d.ts#L800)
