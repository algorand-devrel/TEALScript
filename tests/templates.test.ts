/* eslint-disable no-plusplus */
/* eslint-disable no-empty */
/* eslint-disable no-bitwise */
/* eslint-disable func-names */
/* eslint-disable prefer-arrow-callback */

import * as algokit from '@algorandfoundation/algokit-utils';
import { describe, test, expect } from '@jest/globals';
import { artifactsTest, compileAndCreate, runMethod, algodClient, kmdClient } from './common';
import arc56 from './contracts/artifacts/Templates.arc56.json';

const NAME = 'Templates';
const PATH = 'tests/contracts/general.algo.ts';
const ARTIFACTS_DIR = 'tests/contracts/artifacts/';

const BYTE_CBLOCK = 38;
const INT_CBLOCK = 32;

function getConstantBlockOffsets(bytes: number[]) {
  const programSize = bytes.length;
  bytes.shift(); // remove version
  const offsets: { bytecblockOffset?: number; intcblockOffset?: number; cblocksOffset: number } = { cblocksOffset: 0 };

  while (bytes.length > 0) {
    const byte = bytes.shift()!;

    if (byte === BYTE_CBLOCK || byte === INT_CBLOCK) {
      const isBytecblock = byte === BYTE_CBLOCK;
      const valuesRemaining = bytes.shift()!;

      for (let i = 0; i < valuesRemaining; i++) {
        if (isBytecblock) {
          // byte is the length of the next element
          bytes.splice(0, bytes.shift()!);
        } else {
          // intcblock is a uvarint, so we need to keep reading until we find the end (MSB is not set)
          while ((bytes.shift()! & 0x80) !== 0) {}
        }
      }

      offsets[isBytecblock ? 'bytecblock' : 'intcblock'] = programSize - bytes.length - 1;

      if (bytes[0] !== BYTE_CBLOCK && bytes[0] !== INT_CBLOCK) {
        // if the next opcode isn't a constant block, we're done
        break;
      }
    }
  }

  offsets.cblocksOffset = Math.max(...Object.values(offsets));
  return offsets;
}

describe('Template Variables', function () {
  artifactsTest(PATH, ARTIFACTS_DIR, NAME, {
    templateVars: {
      bytes64TmplVar: `0x${'0'.repeat(64)}`,
      uint64TmplVar: 123,
      bytes32TmplVar: `0x${'0'.repeat(32)}`,
      bytesTmplVar: '0xFF',
    },
  });

  describe('E2E', function () {
    const sender = algokit.getLocalNetDispenserAccount(algodClient, kmdClient);

    test('error source mapping', async function () {
      const { appClient, appId } = await compileAndCreate(await sender, PATH, ARTIFACTS_DIR, NAME, {
        bytes64TmplVar: '0'.repeat(64),
        uint64TmplVar: 123,
        bytes32TmplVar: '0'.repeat(32),
        bytesTmplVar: 'foo',
      });
      let pc = 0;
      try {
        await runMethod({ appClient, method: 'throwError' });
      } catch (error) {
        pc = Number(String(error).match(/pc=(\d+)/)?.[1]);
      }

      const appInfo = await algodClient.getApplicationByID(Number(appId)).do();
      const b64Program = appInfo.params['approval-program'];
      const program = Buffer.from(b64Program, 'base64');
      const { cblocksOffset } = getConstantBlockOffsets([...program]);

      const offsetPc = pc - cblocksOffset;

      const sourceInfo = arc56.sourceInfo.approval.sourceInfo.find((s) => s.pc?.includes(offsetPc));
      expect(sourceInfo?.errorMessage).toBe('this is an error');
    });
  });
});
