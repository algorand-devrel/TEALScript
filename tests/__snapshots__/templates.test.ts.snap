// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Template Variables Templates Artifacts Generates TEAL 1`] = `
"#pragma version 10
intcblock 1 TMPL_uint64TmplVar
bytecblock TMPL_bytesTmplVar TMPL_bytes64TmplVar TMPL_bytes32TmplVar

// https://github.com/algorandfoundation/TEALScript

// This contract is compliant with and/or implements the following ARCs: [ ARC4 ]

// The following ten lines of TEAL handle initial program flow
// This pattern is used to make it easy for anyone to parse the start of the program and determine if a specific action is allowed
// Here, action refers to the OnComplete in combination with whether the app is being created or called
// Every possible action for this contract is represented in the switch statement
// If the action is not implemented in the contract, its respective branch will be "*NOT_IMPLEMENTED" which just contains "err"
txn ApplicationID
!
pushint 6
*
txn OnCompletion
+
switch *call_NoOp *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED *create_NoOp *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED *NOT_IMPLEMENTED

*NOT_IMPLEMENTED:
	// The requested action is not implemented in this contract. Are you using the correct OnComplete? Did you set your app ID?
	err

// tmpl()void
*abi_route_tmpl:
	// execute tmpl()void
	callsub tmpl
	intc 0 // 1
	return

// tmpl(): void
tmpl:
	proto 0 0

	// tests/contracts/general.algo.ts:34
	// log(this.bytesTmplVar)
	bytec 0 // TMPL_bytesTmplVar
	log

	// tests/contracts/general.algo.ts:35
	// assert(this.uint64TmplVar)
	intc 1 // TMPL_uint64TmplVar
	assert
	retsub

// specificLengthTemplateVar()void
*abi_route_specificLengthTemplateVar:
	// execute specificLengthTemplateVar()void
	callsub specificLengthTemplateVar
	intc 0 // 1
	return

// specificLengthTemplateVar(): void
specificLengthTemplateVar:
	proto 0 0

	// tests/contracts/general.algo.ts:39
	// ed25519VerifyBare(this.bytesTmplVar, this.bytes64TmplVar, this.bytes32TmplVar)
	bytec 0 // TMPL_bytesTmplVar
	bytec 1 // TMPL_bytes64TmplVar
	bytec 2 // TMPL_bytes32TmplVar
	ed25519verify_bare
	retsub

// throwError()void
*abi_route_throwError:
	// execute throwError()void
	callsub throwError
	intc 0 // 1
	return

// throwError(): void
throwError:
	proto 0 0

	// this is an error
	err
	retsub

// itobTemplateVar()byte[]
*abi_route_itobTemplateVar:
	// The ABI return prefix
	pushbytes 0x151f7c75

	// execute itobTemplateVar()byte[]
	callsub itobTemplateVar
	dup
	len
	itob
	extract 6 2
	swap
	concat
	concat
	log
	intc 0 // 1
	return

// itobTemplateVar(): bytes
itobTemplateVar:
	proto 0 1

	// tests/contracts/general.algo.ts:47
	// return itob(this.uint64TmplVar);
	intc 1 // TMPL_uint64TmplVar
	itob
	retsub

*abi_route_createApplication:
	intc 0 // 1
	return

*create_NoOp:
	pushbytes 0xb8447b36 // method "createApplication()void"
	txna ApplicationArgs 0
	match *abi_route_createApplication

	// this contract does not implement the given ABI method for create NoOp
	err

*call_NoOp:
	pushbytes 0x9a71d2b4 // method "tmpl()void"
	pushbytes 0xdf4d5c3b // method "specificLengthTemplateVar()void"
	pushbytes 0x3d870d87 // method "throwError()void"
	pushbytes 0xbc0b1706 // method "itobTemplateVar()byte[]"
	txna ApplicationArgs 0
	match *abi_route_tmpl *abi_route_specificLengthTemplateVar *abi_route_throwError *abi_route_itobTemplateVar

	// this contract does not implement the given ABI method for call NoOp
	err"
`;

exports[`Template Variables Templates Artifacts Maintains program size 1`] = `259`;

exports[`Template Variables Templates Artifacts Generates ABI JSON 1`] = `
{
  "desc": "",
  "events": undefined,
  "methods": [
    {
      "args": [],
      "desc": undefined,
      "events": undefined,
      "name": "tmpl",
      "readonly": undefined,
      "returns": {
        "desc": undefined,
        "type": "void",
      },
    },
    {
      "args": [],
      "desc": undefined,
      "events": undefined,
      "name": "specificLengthTemplateVar",
      "readonly": undefined,
      "returns": {
        "desc": undefined,
        "type": "void",
      },
    },
    {
      "args": [],
      "desc": undefined,
      "events": undefined,
      "name": "throwError",
      "readonly": undefined,
      "returns": {
        "desc": undefined,
        "type": "void",
      },
    },
    {
      "args": [],
      "desc": undefined,
      "events": undefined,
      "name": "itobTemplateVar",
      "readonly": undefined,
      "returns": {
        "desc": undefined,
        "type": "byte[]",
      },
    },
    {
      "args": [],
      "desc": undefined,
      "events": undefined,
      "name": "createApplication",
      "readonly": undefined,
      "returns": {
        "desc": undefined,
        "type": "void",
      },
    },
  ],
  "name": "Templates",
}
`;

exports[`Template Variables Templates Artifacts Generates ARC32 1`] = `
{
  "bare_call_config": {
    "close_out": "NEVER",
    "delete_application": "NEVER",
    "no_op": "NEVER",
    "opt_in": "NEVER",
    "update_application": "NEVER",
  },
  "contract": {
    "desc": "",
    "events": undefined,
    "methods": [
      {
        "args": [],
        "desc": undefined,
        "events": undefined,
        "name": "tmpl",
        "readonly": undefined,
        "returns": {
          "desc": undefined,
          "type": "void",
        },
      },
      {
        "args": [],
        "desc": undefined,
        "events": undefined,
        "name": "specificLengthTemplateVar",
        "readonly": undefined,
        "returns": {
          "desc": undefined,
          "type": "void",
        },
      },
      {
        "args": [],
        "desc": undefined,
        "events": undefined,
        "name": "throwError",
        "readonly": undefined,
        "returns": {
          "desc": undefined,
          "type": "void",
        },
      },
      {
        "args": [],
        "desc": undefined,
        "events": undefined,
        "name": "itobTemplateVar",
        "readonly": undefined,
        "returns": {
          "desc": undefined,
          "type": "byte[]",
        },
      },
      {
        "args": [],
        "desc": undefined,
        "events": undefined,
        "name": "createApplication",
        "readonly": undefined,
        "returns": {
          "desc": undefined,
          "type": "void",
        },
      },
    ],
    "name": "Templates",
  },
  "hints": {
    "createApplication()void": {
      "call_config": {
        "no_op": "CREATE",
      },
    },
    "itobTemplateVar()byte[]": {
      "call_config": {
        "no_op": "CALL",
      },
    },
    "specificLengthTemplateVar()void": {
      "call_config": {
        "no_op": "CALL",
      },
    },
    "throwError()void": {
      "call_config": {
        "no_op": "CALL",
      },
    },
    "tmpl()void": {
      "call_config": {
        "no_op": "CALL",
      },
    },
  },
  "schema": {
    "global": {
      "declared": {},
      "reserved": {},
    },
    "local": {
      "declared": {},
      "reserved": {},
    },
  },
  "source": {
    "approval": "I3ByYWdtYSB2ZXJzaW9uIDEwCmludGNibG9jayAxIFRNUExfdWludDY0VG1wbFZhcgpieXRlY2Jsb2NrIFRNUExfYnl0ZXNUbXBsVmFyIFRNUExfYnl0ZXM2NFRtcGxWYXIgVE1QTF9ieXRlczMyVG1wbFZhcgoKLy8gaHR0cHM6Ly9naXRodWIuY29tL2FsZ29yYW5kZm91bmRhdGlvbi9URUFMU2NyaXB0CgovLyBUaGlzIGNvbnRyYWN0IGlzIGNvbXBsaWFudCB3aXRoIGFuZC9vciBpbXBsZW1lbnRzIHRoZSBmb2xsb3dpbmcgQVJDczogWyBBUkM0IF0KCi8vIFRoZSBmb2xsb3dpbmcgdGVuIGxpbmVzIG9mIFRFQUwgaGFuZGxlIGluaXRpYWwgcHJvZ3JhbSBmbG93Ci8vIFRoaXMgcGF0dGVybiBpcyB1c2VkIHRvIG1ha2UgaXQgZWFzeSBmb3IgYW55b25lIHRvIHBhcnNlIHRoZSBzdGFydCBvZiB0aGUgcHJvZ3JhbSBhbmQgZGV0ZXJtaW5lIGlmIGEgc3BlY2lmaWMgYWN0aW9uIGlzIGFsbG93ZWQKLy8gSGVyZSwgYWN0aW9uIHJlZmVycyB0byB0aGUgT25Db21wbGV0ZSBpbiBjb21iaW5hdGlvbiB3aXRoIHdoZXRoZXIgdGhlIGFwcCBpcyBiZWluZyBjcmVhdGVkIG9yIGNhbGxlZAovLyBFdmVyeSBwb3NzaWJsZSBhY3Rpb24gZm9yIHRoaXMgY29udHJhY3QgaXMgcmVwcmVzZW50ZWQgaW4gdGhlIHN3aXRjaCBzdGF0ZW1lbnQKLy8gSWYgdGhlIGFjdGlvbiBpcyBub3QgaW1wbGVtZW50ZWQgaW4gdGhlIGNvbnRyYWN0LCBpdHMgcmVzcGVjdGl2ZSBicmFuY2ggd2lsbCBiZSAiKk5PVF9JTVBMRU1FTlRFRCIgd2hpY2gganVzdCBjb250YWlucyAiZXJyIgp0eG4gQXBwbGljYXRpb25JRAohCnB1c2hpbnQgNgoqCnR4biBPbkNvbXBsZXRpb24KKwpzd2l0Y2ggKmNhbGxfTm9PcCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQgKk5PVF9JTVBMRU1FTlRFRCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQgKmNyZWF0ZV9Ob09wICpOT1RfSU1QTEVNRU5URUQgKk5PVF9JTVBMRU1FTlRFRCAqTk9UX0lNUExFTUVOVEVEICpOT1RfSU1QTEVNRU5URUQgKk5PVF9JTVBMRU1FTlRFRAoKKk5PVF9JTVBMRU1FTlRFRDoKCS8vIFRoZSByZXF1ZXN0ZWQgYWN0aW9uIGlzIG5vdCBpbXBsZW1lbnRlZCBpbiB0aGlzIGNvbnRyYWN0LiBBcmUgeW91IHVzaW5nIHRoZSBjb3JyZWN0IE9uQ29tcGxldGU/IERpZCB5b3Ugc2V0IHlvdXIgYXBwIElEPwoJZXJyCgovLyB0bXBsKCl2b2lkCiphYmlfcm91dGVfdG1wbDoKCS8vIGV4ZWN1dGUgdG1wbCgpdm9pZAoJY2FsbHN1YiB0bXBsCglpbnRjIDAgLy8gMQoJcmV0dXJuCgovLyB0bXBsKCk6IHZvaWQKdG1wbDoKCXByb3RvIDAgMAoKCS8vIHRlc3RzL2NvbnRyYWN0cy9nZW5lcmFsLmFsZ28udHM6MzQKCS8vIGxvZyh0aGlzLmJ5dGVzVG1wbFZhcikKCWJ5dGVjIDAgLy8gVE1QTF9ieXRlc1RtcGxWYXIKCWxvZwoKCS8vIHRlc3RzL2NvbnRyYWN0cy9nZW5lcmFsLmFsZ28udHM6MzUKCS8vIGFzc2VydCh0aGlzLnVpbnQ2NFRtcGxWYXIpCglpbnRjIDEgLy8gVE1QTF91aW50NjRUbXBsVmFyCglhc3NlcnQKCXJldHN1YgoKLy8gc3BlY2lmaWNMZW5ndGhUZW1wbGF0ZVZhcigpdm9pZAoqYWJpX3JvdXRlX3NwZWNpZmljTGVuZ3RoVGVtcGxhdGVWYXI6CgkvLyBleGVjdXRlIHNwZWNpZmljTGVuZ3RoVGVtcGxhdGVWYXIoKXZvaWQKCWNhbGxzdWIgc3BlY2lmaWNMZW5ndGhUZW1wbGF0ZVZhcgoJaW50YyAwIC8vIDEKCXJldHVybgoKLy8gc3BlY2lmaWNMZW5ndGhUZW1wbGF0ZVZhcigpOiB2b2lkCnNwZWNpZmljTGVuZ3RoVGVtcGxhdGVWYXI6Cglwcm90byAwIDAKCgkvLyB0ZXN0cy9jb250cmFjdHMvZ2VuZXJhbC5hbGdvLnRzOjM5CgkvLyBlZDI1NTE5VmVyaWZ5QmFyZSh0aGlzLmJ5dGVzVG1wbFZhciwgdGhpcy5ieXRlczY0VG1wbFZhciwgdGhpcy5ieXRlczMyVG1wbFZhcikKCWJ5dGVjIDAgLy8gVE1QTF9ieXRlc1RtcGxWYXIKCWJ5dGVjIDEgLy8gVE1QTF9ieXRlczY0VG1wbFZhcgoJYnl0ZWMgMiAvLyBUTVBMX2J5dGVzMzJUbXBsVmFyCgllZDI1NTE5dmVyaWZ5X2JhcmUKCXJldHN1YgoKLy8gdGhyb3dFcnJvcigpdm9pZAoqYWJpX3JvdXRlX3Rocm93RXJyb3I6CgkvLyBleGVjdXRlIHRocm93RXJyb3IoKXZvaWQKCWNhbGxzdWIgdGhyb3dFcnJvcgoJaW50YyAwIC8vIDEKCXJldHVybgoKLy8gdGhyb3dFcnJvcigpOiB2b2lkCnRocm93RXJyb3I6Cglwcm90byAwIDAKCgkvLyB0aGlzIGlzIGFuIGVycm9yCgllcnIKCXJldHN1YgoKLy8gaXRvYlRlbXBsYXRlVmFyKClieXRlW10KKmFiaV9yb3V0ZV9pdG9iVGVtcGxhdGVWYXI6CgkvLyBUaGUgQUJJIHJldHVybiBwcmVmaXgKCXB1c2hieXRlcyAweDE1MWY3Yzc1CgoJLy8gZXhlY3V0ZSBpdG9iVGVtcGxhdGVWYXIoKWJ5dGVbXQoJY2FsbHN1YiBpdG9iVGVtcGxhdGVWYXIKCWR1cAoJbGVuCglpdG9iCglleHRyYWN0IDYgMgoJc3dhcAoJY29uY2F0Cgljb25jYXQKCWxvZwoJaW50YyAwIC8vIDEKCXJldHVybgoKLy8gaXRvYlRlbXBsYXRlVmFyKCk6IGJ5dGVzCml0b2JUZW1wbGF0ZVZhcjoKCXByb3RvIDAgMQoKCS8vIHRlc3RzL2NvbnRyYWN0cy9nZW5lcmFsLmFsZ28udHM6NDcKCS8vIHJldHVybiBpdG9iKHRoaXMudWludDY0VG1wbFZhcik7CglpbnRjIDEgLy8gVE1QTF91aW50NjRUbXBsVmFyCglpdG9iCglyZXRzdWIKCiphYmlfcm91dGVfY3JlYXRlQXBwbGljYXRpb246CglpbnRjIDAgLy8gMQoJcmV0dXJuCgoqY3JlYXRlX05vT3A6CglwdXNoYnl0ZXMgMHhiODQ0N2IzNiAvLyBtZXRob2QgImNyZWF0ZUFwcGxpY2F0aW9uKCl2b2lkIgoJdHhuYSBBcHBsaWNhdGlvbkFyZ3MgMAoJbWF0Y2ggKmFiaV9yb3V0ZV9jcmVhdGVBcHBsaWNhdGlvbgoKCS8vIHRoaXMgY29udHJhY3QgZG9lcyBub3QgaW1wbGVtZW50IHRoZSBnaXZlbiBBQkkgbWV0aG9kIGZvciBjcmVhdGUgTm9PcAoJZXJyCgoqY2FsbF9Ob09wOgoJcHVzaGJ5dGVzIDB4OWE3MWQyYjQgLy8gbWV0aG9kICJ0bXBsKCl2b2lkIgoJcHVzaGJ5dGVzIDB4ZGY0ZDVjM2IgLy8gbWV0aG9kICJzcGVjaWZpY0xlbmd0aFRlbXBsYXRlVmFyKCl2b2lkIgoJcHVzaGJ5dGVzIDB4M2Q4NzBkODcgLy8gbWV0aG9kICJ0aHJvd0Vycm9yKCl2b2lkIgoJcHVzaGJ5dGVzIDB4YmMwYjE3MDYgLy8gbWV0aG9kICJpdG9iVGVtcGxhdGVWYXIoKWJ5dGVbXSIKCXR4bmEgQXBwbGljYXRpb25BcmdzIDAKCW1hdGNoICphYmlfcm91dGVfdG1wbCAqYWJpX3JvdXRlX3NwZWNpZmljTGVuZ3RoVGVtcGxhdGVWYXIgKmFiaV9yb3V0ZV90aHJvd0Vycm9yICphYmlfcm91dGVfaXRvYlRlbXBsYXRlVmFyCgoJLy8gdGhpcyBjb250cmFjdCBkb2VzIG5vdCBpbXBsZW1lbnQgdGhlIGdpdmVuIEFCSSBtZXRob2QgZm9yIGNhbGwgTm9PcAoJZXJy",
    "clear": "I3ByYWdtYSB2ZXJzaW9uIDEw",
  },
  "state": {
    "global": {
      "num_byte_slices": 0,
      "num_uints": 0,
    },
    "local": {
      "num_byte_slices": 0,
      "num_uints": 0,
    },
  },
}
`;
